content/blog/improved-kubernetes-await-logic/index.md

```diff
@@ -0,0 +1,293 @@
+
+* More helpful errors are surfaced when a resource doesn't become ready in time.
+* A new `pulumi.com/deletionPropagationPolicy` annotation enables faster deletions.
+* A new `pulumi.com/waitFor` annotation allows for custom readiness criteria.
+* And a new experimental feature enables readiness criteria for all resources.
+
+<!--more-->
+
+## Better error reporting
+
+Our goal is for the Pulumi Kubernetes provider to surface as much information as possible when something fails, because we don't want you to have to manually query your cluster just to figure out what happened.
+
+To that end we've made three improvements to the information Pulumi surfaces when something goes wrong while it's waiting for a resource to become ready (or be deleted):
+
+1. Any Kubernetes Events involving the resource with `type: Warning` are now surfaced by Pulumi as warnings. This can help you see if a Service had trouble provisioning a LoadBalancer, for example.
+
+2. Pulumi will now warn you if it looks like a resource's deletion might be blocked by stuck finalizers. This is a common gotcha that is not always obvious.
+
+3. Error messages from resources that actuate Pods (i.e. Deployments, StatefulSets, etc.) now include status codes and termination messages if their Pods failed to start.
+   A Pod's [termination message](https://kubernetes.io/docs/tasks/debug/debug-application/determine-reason-pod-failure/) can be helpful in `CrashLoopBackOff` situations and is read from `/dev/termination-log` by default.
+
+   (You can configure this path with `terminationMessagePath`, or use `terminationMessagePolicy: FallbackToLogsOnError` to use a Pod's logs as its termination message.)
+
+## Faster deletions with "pulumi.com/deletionPropagationPolicy"
+
+Pulumi uses [foreground](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#foreground-deletion) cascading deletion by default, which deletes a resource _and_ all of its dependents.
+This is the most robust form of deletion but it is also the slowest, and until now there was no way to safely customize this behavior.
+
+The new `pulumi.com/deletionPropagationPolicy` annotation allows you to customize delete behavior while still allowing Pulumi to await readiness when creating or updating the resource.
+
+The annotation accepts three propagation policies:
+
+1. "background": delete the owner resource and leave dependent resources to be asynchronously garbage collected.
+   This is faster than "foreground" deletion propagation, but dependent resources can remain temporarily or even indefinitely if they are not finalized.
+2. "orphan": delete the owner resource and leave dependent resources untouched.
+   This can be useful if you want to keep resources around for migration or debugging purposes.
+3. "foreground": the default behavior of deleting the resource and all of its dependents.
+   This is slower but guarantees all dependents have been cleaned up if it succeeds.
```

Comment:
This exposition makes perfect sense from a Kubernetes perspective, but might confuse when taken in a Pulumi context. Consider making the descriptions be more Kubernetes-centric, like use the term "object".